### PR TITLE
[ST] Multiple cluster operators ST, adding cluster operator metric checks

### PR DESCRIPTION
**Pull request by henryZrncik**
### Type of change

- Refactoring

### Description

Documentation of tests
Addition of metric scrapers and extra asserts regarding which operator operates which operand. 
Refactor to use TestStorage instead of local variables with names.  

### Checklist
Test sometimes fails on diffent problem connected to Cruise Control (but this happens even without this refactor). 

running correctly with FeatureGates (kraft and StableConnectIdentities) and without both Minikube and Openstack.   

- [x] Write tests
- [x] Make sure all tests pass
- [x] Update documentation
- [x] Check RBAC rights for Kubernetes / OpenShift roles

### CODE REVIEW

File: systemtest/src/main/java/io/strimzi/systemtest/utils/specific/MetricsUtils.java [hunk 1]
```diff
@@ -99,14 +100,27 @@ public static String getResourceMetricPattern(String kind, String namespace) {
     }
 
     public static void assertMetricResourcesHigherThanOrEqualTo(MetricsCollector collector, String kind, int expectedValue) {
+        Predicate<Double> higherOrEqualToExpected = actual -> actual >= expectedValue;
```
Comment (see-quick) [inline systemtest/src/main/java/io/strimzi/systemtest/utils/specific/MetricsUtils.java:103]: Just curious...wouldn't be better to factor out this local Predicate to the class scope and also implement an equal Predicate? Therefore you can negate the condition of the following method (i.e., `assertMetricResourcesLowerThanOrEqualTo`) with an equal predicate. WDYT?

**Review by see-quick — APPROVED** [review]

File: systemtest/src/main/java/io/strimzi/systemtest/utils/specific/MetricsUtils.java (hunk 1 above)
Comment (henryZrncik) [reply systemtest/src/main/java/io/strimzi/systemtest/utils/specific/MetricsUtils.java:103]: they are checking 
>= 
and 
<= 

which are not exactly negations.

Comment (henryZrncik) [reply systemtest/src/main/java/io/strimzi/systemtest/utils/specific/MetricsUtils.java:103]: as `=>` and `<=` are not exactly negations I would have to provide additional implementations of another predicate which would have to be in OR relation with negation of another one which may be confusing to read/mantain.

File: systemtest/src/test/java/io/strimzi/systemtest/operators/MultipleClusterOperatorsIsolatedST.java [hunk 2]
```diff
@@ -82,119 +89,211 @@ public class MultipleClusterOperatorsIsolatedST extends AbstractST {
     public static final Map<String, String> FIRST_CO_SELECTOR = Collections.singletonMap("app.kubernetes.io/operator", FIRST_CO_NAME);
     public static final Map<String, String> SECOND_CO_SELECTOR = Collections.singletonMap("app.kubernetes.io/operator", SECOND_CO_NAME);
 
+    /**
+     * @description This test case checks how two Cluster Operators operates operands deployed in namespace watched by both of these operators, and transition
+     * of operand from one Cluster Operator to another.
+     *
+     * @steps
+     *  1. - Deploy all namespace watching first Cluster Operator to the first namespace and second Cluster Operator to the second namespace.
```
Comment (im-konge) [inline systemtest/src/test/java/io/strimzi/systemtest/operators/MultipleClusterOperatorsIsolatedST.java:97]: Maybe?
```suggestion
     *  1. - Deploy two Cluster Operators - first to the first namespace and second to the second namespace - both watching all namespaces.
```

File: systemtest/src/test/java/io/strimzi/systemtest/operators/MultipleClusterOperatorsIsolatedST.java [hunk 3]
```diff
@@ -82,119 +89,211 @@ public class MultipleClusterOperatorsIsolatedST extends AbstractST {
...
+     *  3. - Create and Set namespace 'multiple-co-cluster-test' as default, because all resources from now on will be deployed here.
+     *     - Namespace is created and set as default.
+     *  4. - Deploy Kafka operand in default namespace without label 'app.kubernetes.io/operator'.
+     *     - Kafka is not deployed.
+     *  5. - Verify state of metric 'strimzi_resource' for Kafka, which holds number of operated Kafka operands.
+     *     - Metric 'strimzi_resource' exposed by Cluster Operators is not present indicating no Kafka resource is being operated.
+     *  6. - Modify Kafka custom resource, by changing its label 'app.kubernetes.io/operator' to point to the first Cluster Operator
+     *     - Kafka is deployed and operated by the first Cluster Operator.
+     *  7. - Deploy Kafka Connect (with label 'app.kubernetes.io/operator' pointing the first Cluster Operator) and  Kafka Connector
+     *     - Both Operands are successfully deployed, and managed by the first Cluster Operator.
```
Comment (im-konge) [inline systemtest/src/test/java/io/strimzi/systemtest/operators/MultipleClusterOperatorsIsolatedST.java:110]:
```suggestion
     *     - Both operands are successfully deployed, and managed by the first Cluster Operator.
```

File: systemtest/src/test/java/io/strimzi/systemtest/operators/MultipleClusterOperatorsIsolatedST.java [hunk 4]
```diff
@@ -82,119 +89,211 @@ public class MultipleClusterOperatorsIsolatedST extends AbstractST {
...
+     *  8. - Produce messages into the kafka Topic and consume with Sink Connector.
```
Comment (im-konge) [inline systemtest/src/test/java/io/strimzi/systemtest/operators/MultipleClusterOperatorsIsolatedST.java:111]:
```suggestion
     *  8. - Produce messages into the Kafka Topic and consume with Sink Connector.
```

File: systemtest/src/test/java/io/strimzi/systemtest/operators/MultipleClusterOperatorsIsolatedST.java [hunk 5]
```diff
@@ -205,22 +304,43 @@ void testKafkaCCAndRebalanceWithMultipleCOs(ExtensionContext extensionContext) {
...
+
+        MetricsCollector clusterOperatorCollector = new MetricsCollector.Builder()
+            .withScraperPodName(coScraperPodName)
+            .withNamespaceName(coNamespace)
+            .withComponentType(ComponentType.ClusterOperator)
+            .withComponentName("")
+            .build();
 
-        KafkaRebalanceUtils.doRebalancingProcess(new Reconciliation("test", KafkaRebalance.RESOURCE_KIND, DEFAULT_NAMESPACE, clusterName), DEFAULT_NAMESPACE, clusterName);
+        return clusterOperatorCollector;
```
Comment (im-konge) [inline systemtest/src/test/java/io/strimzi/systemtest/operators/MultipleClusterOperatorsIsolatedST.java:343]: Wouldn't be better to move it to the `MetricsCollector` or some metrics utils class, pass the scraper name as a parameter, so it can be used in the future by other tests? Just an idea, WDYT?

File: systemtest/src/test/java/io/strimzi/systemtest/operators/MultipleClusterOperatorsIsolatedST.java [hunk 6]
```diff
@@ -82,119 +89,211 @@ public class MultipleClusterOperatorsIsolatedST extends AbstractST {
...
+            kafka.getMetadata().getLabels().replace("app.kubernetes.io/operator", SECOND_CO_NAME);
+        }, testStorage.getNamespaceName());
+
+        LOGGER.info("Verifying that number of managed Kafka resources in increased in CO: {} and decreased om CO: {}", SECOND_CO_NAME, FIRST_CO_NAME);
+        MetricsUtils.assertMetricResourcesHigherThanOrEqualTo(secondCoMetricsCollector, Kafka.RESOURCE_KIND, 1);
+        MetricsUtils.assertMetricResourcesLowerThanOrEqualTo(firstCoMetricsCollector, Kafka.RESOURCE_KIND, 0);
     }
 
+    /**
+     * @description This test case checks how two Cluster Operators deployed in the same namespace operates operands including rebalances and transition
```
Comment (im-konge) [inline systemtest/src/test/java/io/strimzi/systemtest/operators/MultipleClusterOperatorsIsolatedST.java:226]:
```suggestion
     * @description This test case checks how two Cluster Operators deployed in the same namespace operates operands including KafkaRebalance and transition
```
maybe?

File: systemtest/src/test/java/io/strimzi/systemtest/operators/MultipleClusterOperatorsIsolatedST.java [hunk 7]
```diff
@@ -82,119 +89,211 @@ public class MultipleClusterOperatorsIsolatedST extends AbstractST {
...
+     * of operand from one Cluster Operator to another.
+     *
+     * @steps
+     *  1. - Deploy 2 Cluster Operators in the same namespace, with additional env variable 'STRIMZI_LEADER_ELECTION_LEASE_NAME'.
+     *     - Cluster Operators are successfully deployed.
+     *  2. - Set up scrapers and metric collectors for first Cluster Operators.
+     *     - Mentioned resources are successfully set up.
+     *  3. - Deploy Kafka Cluster with 3 kafka replicas and label 'app.kubernetes.io/operator' pointing to the first Cluster Operator.
```
Comment (im-konge) [inline systemtest/src/test/java/io/strimzi/systemtest/operators/MultipleClusterOperatorsIsolatedST.java:234]:
```suggestion
     *  3. - Deploy Kafka Cluster with 3 Kafka replicas and label 'app.kubernetes.io/operator' pointing to the first Cluster Operator.
```

**Review by im-konge — COMMENTED** [review]
Just few nits. Anyway, good job :)

**Comment by henryZrncik** [comment]
@strimzi-ci run tests --testcase=MultipleClusterOperatorsIsolatedST

File: systemtest/src/test/java/io/strimzi/systemtest/operators/MultipleClusterOperatorsIsolatedST.java (hunk 5 above)
Comment (henryZrncik) [reply systemtest/src/test/java/io/strimzi/systemtest/operators/MultipleClusterOperatorsIsolatedST.java]: moved to utils. 👍 but i added getting/setting of CO name otherwise this would crash due to not being able to find CO.

File: systemtest/src/main/java/io/strimzi/systemtest/utils/specific/MetricsUtils.java [hunk 8]
```diff
@@ -53,6 +56,25 @@ public static String getExporterRunScript(String podName, String namespace) thro
         return exec.out();
     }
 
+    public static MetricsCollector setupCOMetricsCollectorInNamespace(String coName, String coNamespace, String coScraperName) {
+
+        String originalCODeploymentName = ResourceManager.getCoDeploymentName();
+        // necessary for finding correct deployment when building metric collector.
+        ResourceManager.setCoDeploymentName(coName);
+
+        String coScraperPodName = ResourceManager.kubeClient().listPodsByPrefixInName(coNamespace, coScraperName).get(0).getMetadata().getName();
```
Comment (Frawless) [inline systemtest/src/main/java/io/strimzi/systemtest/utils/specific/MetricsUtils.java:65]: we should list pods by labels

File: systemtest/src/main/java/io/strimzi/systemtest/utils/specific/MetricsUtils.java [hunk 9]
```diff
@@ -53,6 +56,25 @@ public static String getExporterRunScript(String podName, String namespace) thro
...
+
+        MetricsCollector metricsCollector =  new MetricsCollector.Builder()
+            .withScraperPodName(coScraperPodName)
+            .withNamespaceName(coNamespace)
+            .withComponentType(ComponentType.ClusterOperator)
+            .withComponentName("")
```
Comment (Frawless) [inline systemtest/src/main/java/io/strimzi/systemtest/utils/specific/MetricsUtils.java:71]: Maybe not part of this PR, but in case we pass empty string, shouldn't it be a default value?

**Review by Frawless — CHANGES_REQUESTED** [review]

**Comment by henryZrncik** [comment]
@strimzi-ci run tests --testcase=MultipleClusterOperatorsIsolatedST

**Comment by henryZrncik** [comment]
@strimzi-ci run tests --testcase=MultipleClusterOperatorsIsolatedST

File: systemtest/src/main/java/io/strimzi/systemtest/utils/specific/MetricsUtils.java [hunk 10]
```diff
@@ -53,6 +58,19 @@ public static String getExporterRunScript(String podName, String namespace) thro
         return exec.out();
     }
 
+    public static MetricsCollector setupCOMetricsCollectorInNamespace(String coName, String coNamespace, String coScraperName) {
+
+        var scraperDeploymentPodLabel = new LabelSelector(null, Map.of(Constants.APP_POD_LABEL, coScraperName));
```
Comment (im-konge) [inline systemtest/src/main/java/io/strimzi/systemtest/utils/specific/MetricsUtils.java:63]:
```suggestion
        LabelSelector scraperDeploymentPodLabel = new LabelSelector(null, Map.of(Constants.APP_POD_LABEL, coScraperName));
```

File: systemtest/src/main/java/io/strimzi/systemtest/utils/specific/MetricsUtils.java [hunk 11]
```diff
@@ -53,6 +58,19 @@ public static String getExporterRunScript(String podName, String namespace) thro
...
+        String coScraperPodName = ResourceManager.kubeClient().listPods(coNamespace, scraperDeploymentPodLabel).get(0).getMetadata().getName();
+
+        return  new MetricsCollector.Builder()
```
Comment (im-konge) [inline systemtest/src/main/java/io/strimzi/systemtest/utils/specific/MetricsUtils.java:66]:
```suggestion
        return new MetricsCollector.Builder()
```

**Review by im-konge — CHANGES_REQUESTED** [review]

**Review by im-konge — APPROVED** [review]

**Review by Frawless — APPROVED** [review]

**Comment by im-konge** [comment]
The failure is not connected to this PR - #8716 will fix it.
